### PR TITLE
Support absolute url to override base url

### LINKED ISSUE #10027: AssertionError  | assert not url.absolute raisedon a WSS URL
### Describe the bug

Discord's [` Get Gateway `](https://discord.com/developers/docs/events/gateway#get-gateway) endpoint returns a ` url ` field containing ` "wss://gateway.discord.gg" `. This WSS URL is used to establish connection with their gateway.  Though this raises an exception:
```
File "...\aiohttp\client.py", line 467, in _build_url
  assert not url.absolute
         ^^^^^^^^^^^^^^^^^
AssertionError
```
This is also tested with applying ` yarl.URL(wss_url) ` ( maybe ), but same issue.

Side notes:
1. I am testing in both Python 3.12 and 3.13, but would be more favorable to me if there is a fix already for 3.13
2. I may provide other information should you ask relating to it

### To Reproduce

```py
from aiohttp import ClientSession
from asyncio import run
from typing import *

class DiscordWebSocket:
  session = lambda: ClientSession(base_url = "https://discord.com/api/v10/")
  connection = None # would likely be replaced by DiscordWebSocket.connect()

  @classmethod
  async def connect(cls) -> NoReturn:
    wss : Dict = await cls.get("gateway") # {"url": "wss://gateway.discord.gg"}
    async with cls.session() as session:
      response = await session.ws_connect(f"{wss['url']}/") # AssertionError
      return response # debug stuff lol

  @classmethod
  async def get(cls, endpoint : str) -> Dict:
    async with cls.session() as session:
      response = session.get(endpoint)
      return await response.json()


async def main() -> NoReturn:
  print(f"{await DiscordWebSocket.connect() = }")

run(main())
```
1. Retrieve the WSS URL from [` Get Gateway `](https://discord.com/developers/docs/events/gateway#get-gateway) endpoint and pass it to ` async ClientSession.ws_connect() `

### Expected behavior

It *would* ( ? should ? ) print ` await DiscordWebSocket.connect() = <aiohttp.ClientWebSocketResponse ...> ` in the console

### Logs/tracebacks

```python-traceback
> python .\main.py
Traceback (most recent call last):
  File "C:\Users\demo\OneDrive\Documents\python\test\main.py", line 32, in <module>
    run(main())
  File "C:\Users\demo\AppData\Local\Programs\Python\Python312\Lib\asyncio\runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\demo\AppData\Local\Programs\Python\Python312\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\demo\AppData\Local\Programs\Python\Python312\Lib\asyncio\base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "C:\Users\demo\OneDrive\Documents\python\test\main.py", line 29, in main
    print(f"{await WebSocket.connect() = }")
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\demo\OneDrive\Documents\python\test\main.py", line 16, in connect
    response = await session.ws_connect(str(URL(f"{wss['url']}/")))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\demo\AppData\Local\Programs\Python\Python312\Lib\site-packages\aiohttp\client.py", line 1002, in _ws_connect
    resp = await self.request(
           ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\demo\AppData\Local\Programs\Python\Python312\Lib\site-packages\aiohttp\client.py", line 535, in _request
    url = self._build_url(str_or_url)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\demo\AppData\Local\Programs\Python\Python312\Lib\site-packages\aiohttp\client.py", line 467, in _build_url
    assert not url.absolute
           ^^^^^^^^^^^^^^^^
AssertionError
```

### Python Version

```console
$ python --version
Python 3.12.4
Python 3.13.0
```

### aiohttp Version

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.11.7
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author:
Author-email:
License: Apache-2.0
Location: C:\Users\demo\AppData\Local\Programs\Python\Python312\Lib\site-packages
Requires: aiohappyeyeballs, aiosignal, attrs, frozenlist, multidict, propcache, yarl
Required-by: discord.py
```

### multidict Version

```console
$ python -m pip show multidict
Version: 6.0.5
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: C:\Users\demo\AppData\Local\Programs\Python\Python312\Lib\site-packages
Requires:
Required-by: aiohttp, yarl
```

### propcache Version

```console
$ python -m pip show propcache
Name: propcache
Version: 0.2.0
Summary: Accelerated property cache
Home-page: https://github.com/aio-libs/propcache
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: C:\Users\demo\AppData\Local\Programs\Python\Python312\Lib\site-packages
Requires:
Required-by: aiohttp, yarl
```

### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.18.0
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: C:\Users\demo\AppData\Local\Programs\Python\Python312\Lib\site-packages
Requires: idna, multidict, propcache
Required-by: aiohttp
```

### OS

Windows 10

### Related component

Client

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

### PR DESCRIPTION
**Pull request by vivodi**
## What do these changes do?

Support absolute url in request to override base url in session.
```py
session = ClientSession(base_url="http://example.com/foo/")

await session.request("GET", "http://abc.com/bar/")
```

## Are there changes in behavior for the user?

Yes

## Is it a substantial burden for the maintainers to support this?

No

## Related issue number

Fixes #10027
## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES/` folder
  * name it `<issue_or_pr_num>.<type>.rst` (e.g. `588.bugfix.rst`)
  * if you don't have an issue number, change it to the pull request
    number after creating the PR
    * `.bugfix`: A bug fix for something the maintainers deemed an
      improper undesired behavior that got corrected to match
      pre-agreed expectations.
    * `.feature`: A new behavior, public APIs. That sort of stuff.
    * `.deprecation`: A declaration of future API removals and breaking
      changes in behavior.
    * `.breaking`: When something public is removed in a breaking way.
      Could be deprecated in an earlier release.
    * `.doc`: Notable updates to the documentation structure or build
      process.
    * `.packaging`: Notes for downstreams about unobvious side effects
      and tooling. Changes in the test invocation considerations and
      runtime assumptions.
    * `.contrib`: Stuff that affects the contributor experience. e.g.
      Running tests, building the docs, setting up the development
      environment.
    * `.misc`: Changes that are hard to assign to any of the above
      categories.
  * Make sure to use full sentences with correct case and punctuation,
    for example:
    ```rst
    Fixed issue with non-ascii contents in doctest text files
    -- by :user:`contributor-gh-handle`.
    ```

    Use the past tense or the present tense a non-imperative mood,
    referring to what's changed compared to the last released version
    of this project.

### CODE REVIEW

File: CHANGES/10074.feature.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Supported absolute url to override base url.
```
Comment (webknjaz) [inline CHANGES/10074.feature.rst]:
```suggestion
Added support for overriding the base URL with an absolute one in client sessions
```

File: aiohttp/client.py [hunk 2]
```diff
@@ -414,11 +414,9 @@ def request(
 
     def _build_url(self, str_or_url: StrOrURL) -> URL:
         url = URL(str_or_url)
-        if self._base_url is None:
-            return url
-        else:
-            assert not url.absolute
-            return self._base_url.join(url)
+        return (
+            url if self._base_url is None or url.absolute else self._base_url.join(url)
```
Comment (webknjaz) [inline aiohttp/client.py]: Could you put the conditional into a named variable for readability?

File: docs/client_reference.rst [hunk 3]
```diff
@@ -77,6 +77,9 @@ The client session supports the context manager protocol for self closing.
         await session.request("GET", "/bar")
         # request for http://example.com/bar
 
+        await session.request("GET", "http://foo.com/bar")
```
Comment (webknjaz) [inline docs/client_reference.rst]: This wasn't added in 3.8, though. The docs should separate it, I think.

**Review by webknjaz — COMMENTED** [review]
Such a change would require tests to be included.

**Comment by vivodi** [comment]
The PR already meets all your other requirements, but if I put the condition in a named variable, it fails the Linter check:
```py
need_join = self._base_url and not url.absolute
return self._base_url.join(url) if need_join else url
```
Because Linter thinks that `self._base_url` may be `None`, and `None` does not have attribute `join`.
What should I do?

File: aiohttp/client.py [hunk 4]
```diff
@@ -414,11 +414,8 @@ def request(
 
     def _build_url(self, str_or_url: StrOrURL) -> URL:
         url = URL(str_or_url)
-        if self._base_url is None:
-            return url
-        else:
-            assert not url.absolute
-            return self._base_url.join(url)
+        need_join = self._base_url and not url.absolute
+        return self._base_url.join(url) if need_join else url
```
Comment (vivodi) [inline aiohttp/client.py]: The PR already meets all your other requirements, but if I put the condition in a named variable, it fails the Linter check.

Linter thinks that `self._base_url` may be `None`, and `None` does not have attribute `join`.
What should I do?

Comment (webknjaz) [reply aiohttp/client.py]: Let's try
```suggestion
        is_relative_path = self._base_url and not url.absolute
        if is_relative_path:
            return self._base_url.join(url)

        return url
```
then. @Dreamsorcerer mentioned type narrowing being imperfect in MyPy..

Comment (Dreamsorcerer) [reply aiohttp/client.py]: Yeah, that will just result in the variable being a `bool`, it won't narrow the type of `_base_url` in the if block.

Best to drop the variable entirely (avoiding the one-liner should retain some readability):
```suggestion
        if self._base_url and not url.absolute:
            return self._base_url.join(url)
        return url
```

**Review by Dreamsorcerer — APPROVED** [review]

**Review by webknjaz — APPROVED** [review]
LGTM and the docs render well: https://aiohttp--10074.org.readthedocs.build/en/10074/client_reference.html#aiohttp.ClientSession

**Review by bdraco — APPROVED** [review]

**Pull request merged**